### PR TITLE
Add post-fetch filtering to bulk export

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

**Review by jmarchionatto — APPROVED** [review]
Looks good! Approved with a few questions

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_6_0/4772-add-post-fetch-filtering-to-bulk-export.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+type: add
+issue: 4772
+backport: 6.4.5
+title: "A new optional parameter called `_typePostFetchFilterUrl` has been added to the 
+   Bulk Export `$export` 
+   operation parameters. This parameter allows filters to be specified that will be applied
+   in-memory to the resources after that have been initially fetched by the database. This can
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_6_0/4772-add-post-fetch-filtering-to-bulk-export.yaml:8]: thanks, fixed

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/model/BulkExportParameters.java [hunk 2]
```diff
@@ -48,6 +48,8 @@ public class BulkExportParameters extends Batch2BaseJobParameters {
 	 */
 	private List<String> myFilters;
 
+	private List<String> myPostFetchFilterUrls;
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/model/BulkExportParameters.java:51]: added

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/ExpandResourcesStep.java [hunk 3]
```diff
@@ -116,27 +133,62 @@ public RunOutcome run(@Nonnull StepExecutionDetails<BulkExportJobParameters, Res
 		}
 
 		// encode them - Key is resource type, Value is a collection of serialized resources of that type
-		ListMultimap<String, String> resources = encodeToString(allResources, jobParameters);
+		ListMultimap<String, String> resources = encodeToString(allResources, parameters);
 
 		// set to datasink
-		for (String nextResourceType : resources.keySet()) {
+		if (!resources.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/ExpandResourcesStep.java:139]: you know what, you're absolutely right, this does nothing. Removing the if. Good catch!

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/ExpandResourcesStep.java [hunk 4]
```diff
@@ -116,27 +133,62 @@ public RunOutcome run(@Nonnull StepExecutionDetails<BulkExportJobParameters, Res
...
 	}
 
+	private void applyPostFetchFiltering(List<IBaseResource> theResources, List<String> thePostFetchFilterUrls, String theInstanceId, String theChunkId) {
+		int numRemoved = 0;
+		for (Iterator<IBaseResource> iter = theResources.iterator(); iter.hasNext(); ) {
+			IBaseResource nextResource = iter.next();
+			String nextResourceType = myFhirContext.getResourceType(nextResource);
+			boolean matched = false;
+
+			for (String nextPostFetchFilterUrl : thePostFetchFilterUrls) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/ExpandResourcesStep.java:165]: i like it. done.

**Pull request merged**